File: Java/프로그래머스/3/64064. 불량 사용자/불량 사용자.java
import java.util.*;
class Solution {
    public int solution(String[] user_id, String[] banned_id) {
        List<Integer>[] list = new ArrayList[banned_id.length];
        for (int i = 0; i < banned_id.length; i++) {
            list[i] = new ArrayList<>();
        }
        for (int i = 0; i < user_id.length; i++) {
            String user = user_id[i];
            for (int j = 0; j < banned_id.length; j++) {
                String ban = banned_id[j];
                if (isMatch(user, ban)) {
                    list[j].add(i);
                }
            }
        }

        Set<Set<Integer>> set = new HashSet<>();
        dfs(0, list, set, new HashSet<>());

        return set.size();
    }

    private void dfs(int depth, List<Integer>[] list, Set<Set<Integer>> set, Set<Integer> tmp) {
        if (depth == list.length) {
            set.add(new HashSet<>(tmp));
            return;
        }

        for (int user : list[depth]) {
            if (tmp.contains(user)) continue;
            tmp.add(user);
            dfs(depth + 1, list, set, tmp);
            tmp.remove(user);
        }
    }

    private boolean isMatch(String user, String ban) {
        if (user.length() != ban.length()) return false;

        for (int i = 0; i < user.length(); i++) {
            if (user.charAt(i) != ban.charAt(i) && ban.charAt(i) != '*') return false;
        }

        return true;
    }
}

File: Python/프로그래머스/3/64064. 불량 사용자/불량 사용자.py
def solution(user_id, banned_id):
    lst = [[] for _ in range(len(banned_id))]

    for i, user in enumerate(user_id):
        for j, ban in enumerate(banned_id):
            if is_match(user, ban):
                lst[j].append(i)

    st = set()
    dfs(0, lst, st, set())
    return len(st)


def dfs(depth, lst, st, tmp):
    if depth == len(lst):
        st.add(frozenset(tmp))
        return

    for user in lst[depth]:
        if user in tmp:
            continue
        tmp.add(user)
        dfs(depth + 1, lst, st, tmp)
        tmp.remove(user)


def is_match(user, ban):
    if len(user) != len(ban):
        return False

    for i in range(len(user)):
        if user[i] != ban[i] and ban[i] != "*":
            return False

    return True